Define Coordinator and Client

marcoscaceres opened the pull request:
Closes #305

TallTed commented on index.html:
```diff
@@ -285,6 +285,104 @@ <h2>
+    </p>
+    <ul>
+      <li>Ensures that only one [=digital credential=] interaction is active at
+      any time. This includes both via the:
```
```suggestion
      any time. This includes via both the:
```

mohamedamir commented on index.html:
```diff
@@ -285,6 +285,110 @@ <h2>
+          </li>
+          <li>
+            <dfn data-dfn-for=
+            "credential request coordinator">Presenting</dfn>: A credential
```
`Presenting` is very tailored for presentation, but doesn't work well with Issuance.

```suggestion
            "credential request coordinator">Ongoing</dfn>: A credential
```

marcoscaceres replied:
Good call

mohamedamir commented on index.html:
```diff
@@ -285,6 +285,110 @@ <h2>
+      Digital Credential Client
+    </h3>
+    <p>
+      A <dfn>digital credential client</dfn> is a user-agent-defined component
+      responsible for presenting available [=holders=] to the user and
+      coordinating the credential request or issuance flow.
```
I would also avoid using presenting here in that context.

```suggestion
      A <dfn>digital credential client</dfn> is a user-agent-defined component
      responsible for facilitating the selection of available [=holders=] and
      coordinating the credential presentation or issuance flow.
```

mohamedamir replied:
Also another option
```suggestion
      A <dfn>digital credential client</dfn> is a user-agent-defined component
      responsible managing the credential presentation or issuance flow, which 
      includes allowing a user to select from available [=holders=] and coordinating the request.
```

mohamedamir commented on index.html:
```diff
@@ -285,6 +285,110 @@ <h2>
+    </p>
+    <p>
+      Each [=credential request coordinator=] uses a single associated
+      [=digital credential client=] to fulfill credential request and issuance
```
IMHO, `request` is comparable to `create`
and `presentation` is comparable to `issuance`
So `request` and `issuance` isn't that accurate I think.

```suggestion
      [=digital credential client=] to fulfill credential presentation and issuance
```

mohamedamir commented on index.html:
```diff
@@ -285,6 +285,110 @@ <h2>
+      The [=digital credential client=]:
+    </p>
+    <ul>
+      <li>Validates and transforms digital credential request or issuance
```
same here
```suggestion
      <li>Validates and transforms digital credential presentation or issuance
```

mohamedamir commented on index.html:
```diff
@@ -285,6 +285,110 @@ <h2>
+        [=credential chooser=].
+      </p>
+      <p>
+        In this model, the [=digital credential client=] presents the user with
+        a list of available [=holders=], and the selected [=holder=] then
```
Similarly, I don't think it's the client.

marcoscaceres replied:
Yes, also ok to drop this.

marcoscaceres commented on index.html:
```diff
@@ -285,6 +285,110 @@ <h2>
+      </li>
+      <li>Presents the available [=holders=] to the user for selection.
+      </li>
+      <li>Once a [=holder=] is selected, invokes it as a [=credential chooser=]
+      to complete the interaction.
+      </li>
```
```suggestion
```

marcoscaceres commented on index.html:
```diff
@@ -285,6 +285,107 @@ <h2>
+        interaction, it is the selected [=holder=] that typically acts as the
+        [=credential chooser=].
+      </p>
+      <p>
+        In this model, the [=digital credential client=] presents the user with
+        a list of available [=holders=], and the selected [=holder=] then
+        displays its own user interface for selecting or issuing [=digital
+        credentials=].
+      </p>
```
```suggestion
```

mohamedamir commented on index.html:
```diff
@@ -285,6 +285,110 @@ <h2>
+      </li>
+      <li>Once a [=holder=] is selected, invokes it as a [=credential chooser=]
+      to complete the interaction.
+      </li>
```
I still see some of this in the PR.
Is that intentional?

marcoscaceres replied:
No. Would you mind suggesting any additional changes if I missed something? 🙏

TallTed commented on index.html:
```diff
@@ -349,29 +349,26 @@ <h3>
       A <dfn>digital credential client</dfn> is a user-agent-defined component
-      responsible for presenting available [=holders=] to the user and
-      coordinating the credential request or issuance flow.
+      responsible managing the credential presentation or issuance flow, which 
```
```suggestion
      responsible for managing the credential presentation or issuance flow, which 
```

TallTed commented on index.html:
```diff
@@ -283,6 +283,101 @@ <h2>
+      inputs.
+      </li>
+      <li>Interprets each individual request in the input list as targeting a
+      specific [=holder=], and prepares holder options accordingly.
```
```suggestion
      specific [=holder=], and prepares [=holder=] options accordingly.
```

marcoscaceres commented on index.html:
```diff
@@ -283,6 +283,101 @@ <h2>
+      </li>
+      <li>Interprets each individual request in the input list as targeting a
+      specific [=holder=], and prepares holder options accordingly.
+      </li>
```
This shouldn't be WebKit specific, no... the matcher should be generic.

marcoscaceres commented on index.html:
```diff
@@ -283,6 +283,101 @@ <h2>
+      </li>
+      <li>Presents the available [=holders=] to the user for selection.
+      </li>
+      <li>Returns the result — a selected credential or a cancellation — to the
```
```suggestion
      <li>Returns the result — the user's selected [=digital credential=] or a cancellation — to the
```

marcoscaceres commented:
@mohamedamir added the diagram above, as we discussed... will fix it up and add to the spec.

mohamedamir commented:
> @mohamedamir added the diagram above, as we discussed... will fix it up and add to the spec.

Great!
Thanks @marcoscaceres

TallTed commented on index.html:
```diff
@@ -430,7 +430,175 @@ <h2>
+    <p>
+      The [=credential request coordinator=] has a finite set of
+      <dfn data-dfn-for="credential request coordinator">interaction
+      states</dfn>, which are used to manage the lifecycle of request for a
+      digital credential, be it either a [=digital credential/presentation
```
```suggestion
      states</dfn>, which are used to manage the lifecycle of a request for a
      digital credential, whether it is a [=digital credential/presentation
```

TallTed commented on index.html:
```diff
@@ -430,7 +430,175 @@ <h2>
+        <dfn data-dfn-for="credential request coordinator">aborting</dfn>:
+      </dt>
+      <dd>
+        A credential interaction is being canceled, either because of an error,
+        user cancellation, or a [=AbortController/signal abort=]. The
```
```suggestion
        A credential interaction is being canceled, because of an error,
        a user cancellation, or a [=AbortController/signal abort=]. The
```

TallTed commented on index.html:
```diff
@@ -430,7 +430,175 @@ <h2>
+        interface is being presented to the user.
+      </dd>
+      <dt>
+        <dfn data-dfn-for="credential request coordinator">aborting</dfn>:
```
```suggestion
        <dfn data-dfn-for="credential request coordinator">Aborting</dfn>:
```

TallTed commented on index.html:
```diff
@@ -430,7 +430,175 @@ <h2>
+        No credential interaction is currently in progress.
+      </dd>
+      <dt>
+        <dfn data-dfn-for="credential request coordinator">requesting</dfn>:
```
```suggestion
        <dfn data-dfn-for="credential request coordinator">Requesting</dfn>:
```

TallTed commented on index.html:
```diff
@@ -430,7 +430,175 @@ <h2>
+    </p>
+    <dl>
+      <dt>
+        <dfn data-dfn-for="credential request coordinator">idle</dfn>:
```
```suggestion
        <dfn data-dfn-for="credential request coordinator">Idle</dfn>:
```

TallTed commented on index.html:
```diff
@@ -430,7 +430,175 @@ <h2>
+    <ul>
+      <li>Validates and transforms presentation or issuance inputs and outputs.
+      </li>
+      <li>Interprets each individual request in the input list as targeting a
+      specific [=holder=], and prepares [=holder=] options accordingly.
+      </li>
+      <li>Presents the available [=holders=] to the user for selection.
+      </li>
+      <li>Returns the result — the user's selected [=digital credential=] or a
+      cancellation — to the [=credential request coordinator=].
```
```suggestion
      <li>Interprets each individual request in the input list as targeting a
        specific [=holder=], and prepares [=holder=] options accordingly.
      </li>
      <li>Presents the available [=holders=] to the user for selection.
      </li>
      <li>Returns the result — the user's selected [=digital credential=] or a
        cancellation — to the [=credential request coordinator=].
```

TallTed commented on index.html:
```diff
@@ -430,7 +430,175 @@ <h2>
+      The following items are within the scope of this specification:
+    </p>
+    <ul>
+      <li>Requesting a [=digital credential=], including mechanisms for secure
+      presentation.
+      </li>
+      <li>Requesting issuance of a [=digital credential=].
+      </li>
+      <li>Ensuring that when an API call is made, the website does not learn
+      anything about the a holder's [=digital credentials=] or their software
+      unless the [=user agent=] has previously received user consent.
+      </li>
+      <li>Ensuring that any installed application software will not learn
+      anything about a given [=request=] unless the [=holder=] explicitly
+      consents to use that software.
+      </li>
```
```suggestion
      <li>Requesting a [=digital credential=], including mechanisms for secure
        presentation.
      </li>
      <li>Requesting issuance of a [=digital credential=].
      </li>
      <li>Ensuring that when an API call is made, the website does not learn
        anything about the a holder's [=digital credentials=] or their software
        unless the [=user agent=] has previously received user consent to reveal
        that information.
      </li>
      <li>Ensuring that any installed application software will not learn
        anything about a given [=request=] unless the [=holder=] explicitly
        consents to reveal the [=request=] to that software.
      </li>
```

TallTed commented on index.html:
```diff
@@ -430,7 +430,175 @@ <h2>
+    </p>
+    <ul>
+      <li>UI/UX considerations, with the exception of privacy considerations,
+      which are addressed to ensure the protection of user data during the
+      request process.
```
```suggestion
        which are addressed to ensure the protection of user data during the
        request process.
```

TallTed reviewed (changes requested):
Tweaks for consistency within source, thence to consistency in rendering for humans.

hlflanagan commented:
Discussed on the [FedID WG DC API Series B call](https://github.com/w3c-fedid/meetings/blob/main/2025/2025-08-06-DCAPI-B-notes.md)

marcoscaceres commented:
> It is also essentially mandates that the credential selector is a UA component, which is not always the case (in most cases it is not).

Sorry, that is not the intent. It's not a UA component, it's either a platform component or generated by the OS. 

> This is why I previously had the DC Client and DC Client Platform definitions which can clearly separate some of this.

We can just clarify that in the text.

marcoscaceres commented on index.html:
```diff
@@ -449,7 +449,176 @@ <h2>
+    <h2>
+      Credential request client
+    </h2>
+    <p>
+      A <dfn data-local-lt="client">credential request client</dfn> is a
+      user-agent-defined component responsible for managing the credential
+      presentation or issuance flow, which includes allowing the user to select
+      from available [=holders=] and coordinating the request.
+    </p>
```
```suggestion
    <p>
A <dfn data-local-lt="client">credential request client</dfn> is a component responsible for managing the credential presentation or issuance flow, as defined or delegated by the user agent. The credential request client may be implemented by the user agent itself, by an installed wallet, by platform-provided software, or by another trusted component.    
</p>
<aside class="note"> <p>The user agent can delegate some or all credential request client responsibilities to external wallet applications, platform components, or other trusted entities, according to user or platform policy. </p> </aside>
```

marcoscaceres commented:
@timcappalli, fixed up wording to address your concern (clarified how the client might be implemented). 

Added a detailed description at the top that should clarify why both these components are needed by _all user agents_ (these are not UA specific) and how they manage the API's lifecycle across browsing contexts, with respect to the top-level traversable.

TallTed commented:
@marcoscaceres — Good to know whitespace may be fixed by CI scripting. Sadly, relying on such things has bitten me uncountable times in the past, so I'm likely to continue to flag them. Also, I note that in the [`tidy`](https://github.com/w3c-fedid/digital-credentials/pull/306/commits/fc7a2f453d89a9d120061c1987d34a978b20961c) step above, [indents I had suggested](https://github.com/w3c-fedid/digital-credentials/pull/306#pullrequestreview-3094331095) within `<li>` items [*and you had applied*](https://github.com/w3c-fedid/digital-credentials/pull/306/commits/758720432a26dd7767f05bc3da87383dd24e1ebe) are *deleted*, so apparently the `tidy` script doesn't actually comply with best practices.

marcoscaceres commented:
> tidy script doesn't actually comply with best practices. 

I'll check what happened, but whatever tidy does *is* best practice. Last thing we want to do is maintain any whitespace by hand. 

Personally, I've been using tidy across maybe ~50 specs for over a decade and have never had any issues with it.

TallTed commented:
> whatever tidy does _is_ best practice

I would say rather, "whatever tidy does _is meant to be_ best practice". Bugs happen, both in code and in coders' minds, and this appears to be the case here.

hlflanagan commented:
Discussed on the [DC API call - 11 August](https://github.com/w3c-fedid/meetings/blob/main/2025/2025-08-11-DCAPI-A-notes.md). No consensus yet on this item.

marcoscaceres commented:
@mohamedamir, that makes sense, just the naming is still a bit too abstract, and yes it's good to call out the platform. Note that on Apple's Platforms it's the platform that renders the QR code, not the user agent. But I can see that it could be UA that does that maybe in Chrome. We can account for that.

marcoscaceres commented:
On the call splitting this into two PRs.

we can then continue making progress on the boundary of “client” (or whatever we want to call it).

marcoscaceres commented:
Reduced purpose: 

* Handle passing the request to the platform (presents the credential chooser, including origin)
* handle `.abort()` from AbortController
* handle abort from the user - user presses "esc" or hits "X" or whatever via UI.
* Handle "credential got picked" from the platform -> resolve promise. 
* Handle page specific events:
  * Page got navigated away
  * Page (iframe) got destroyed

hlflanagan commented:
Discussed on [25 August 2025 call](https://github.com/w3c-fedid/meetings/blob/main/2025/2025-08-25-DCAPI-A-notes.md#define-coordinator-and-client-306). Agree to changes in principle but need to see the revised text.

TallTed commented on index.html:
```diff
@@ -492,7 +492,7 @@ <h3>
     <ul>
       <li>Ensures that only one [=digital credential=] interaction is active at
-      any time. This includes both via the:
+      any time. This includes via both the:
```
A colon can only be used in this fashion when it could sensibly be replaced by a period (a/k/a "full stop").
```suggestion
      any time, including by use of either or both of the following:
```

TallTed commented on index.html:
```diff
@@ -539,8 +539,8 @@ <h3>
       A <dfn>digital credential client</dfn> is a user-agent-defined component
-      responsible for presenting available [=holders=] to the user and
-      coordinating the credential request or issuance flow.
+      responsible managing the credential presentation or issuance flow, which 
```
```suggestion
      responsible for managing the credential presentation or issuance flow, which 
```

TallTed commented on index.html:
```diff
@@ -471,105 +471,145 @@ <h2>
-      The [=credential request coordinator=]:
+      The coordinator manages the lifecycle of the current interaction's
+      {{Promise}}, including its resolution or rejection and integrates with an
+      {{AbortSignal}} to allow for both user-initiated and programmatic
```
```suggestion
      {{AbortSignal}} to allow for both user-initiated and program-initiated
```

TallTed commented on index.html:
```diff
@@ -471,105 +471,145 @@ <h2>
+        <dfn data-dfn-for="credential request coordinator">aborting</dfn>:
+      </dt>
+      <dd>
+        A credential interaction is being canceled, either because of an error,
```
```suggestion
        A credential interaction is being canceled, because of either an error, a
```

TallTed commented on index.html:
```diff
@@ -471,105 +471,145 @@ <h2>
+      A <dfn data-local-lt="client">credential request client</dfn> is a
+      user-agent-defined component responsible for managing the credential
+      presentation or issuance flow, which includes allowing the user to select
+      from available [=holders=] and coordinating the request.
```
```suggestion
      user-agent-defined component responsible for managing the flow of
      credential presentation or issuance, which includes allowing the user
      to select from available [=holders=] and coordinating the request.
```

TallTed commented on index.html:
```diff
@@ -471,105 +471,145 @@ <h2>
-        [=credential chooser=].
+        Although the [=credential request client=] coordinates the user
+        interaction with respect to inputs and outputs, it is the [=holder=]
+        selected by the user that typically acts as the [=credential chooser=].
```
```suggestion
        Although the [=credential request client=] coordinates the inputs and
        outputs of a user interaction, it is the [=holder=] selected by the
        user that typically acts as the [=credential chooser=].
```

TallTed commented on index.html:
```diff
@@ -491,64 +485,80 @@ <h2>
+      interactions within a [=top-level traversable=]. Each [=top-level
+      traversable=] has exactly one associated coordinator. The coordinator
+      ensures that at most one interaction is active across all [=child
+      navigables=], orchestrates the end-to-end presentation or issuance flow,
```
```suggestion
      navigables=], orchestrates the end-to-end flow of presentation or issuance,
```

TallTed commented on index.html:
```diff
@@ -491,64 +485,80 @@ <h2>
-      cancellation.
+      The coordinator manages the lifecycle of the interaction's {{Promise}}
+      and its associated {{AbortSignal}}, including resolution with the user's
+      selected [=digital credential=] (or credentials) or rejection due to
+      errors or user/programmatic cancellation.
```
selected [=digital credential|digital credential(s)=] or rejection due to
      errors or cancellation by user or program.

TallTed commented on index.html:
```diff
@@ -491,64 +485,80 @@ <h2>
-      states</dfn>, which are used to manage the lifecycle of a request for a
-      digital credential, whether it is a [=digital credential/presentation
-      request=] or an [=digital credential/issuance request=]. The states are:
+      states</dfn> used to manage the lifecycle of a [=digital
```
These two words should not be deleted.
```suggestion
      states</dfn>, which are used to manage the lifecycle of a [=digital
```

TallTed commented on index.html:
```diff
@@ -491,64 +485,80 @@ <h2>
-        cancellation, or a [=AbortController/signal abort=]. The coordinator is
-        in the process of cleaning up the interaction in order to return to the
-        [=credential request coordinator/idle=] state.
+        The active interaction is being canceled due to an error, user action,
+        or an [=AbortController/signal abort=]; the coordinator is cleaning up
+        before returning to [=credential request coordinator/idle=].
```
Either all of the causes of cancellation should have an indefinite article (`a`/`an`), or only the first should. I think all is better. Also, `an` should not precede consonant-sounding words (e.g., `signal`), only vowel-sounding word (e.g., `abort`).
```suggestion
        The active interaction is being canceled due to an error, a user action,
        or a [=AbortController/signal abort=]; the coordinator is cleaning up
        before returning to [=credential request coordinator/idle=].
```

marcoscaceres commented:
Thanks @TallTed! much improved... note that ReSpec doesn't support `[=word(s)=]`... but that could be a cool addition... might add support for that 😍

marcoscaceres commented:
@TallTed, oh! with suggesting alternatives, it's usually just better to pick one or otherwise GitHub seems to get confused. Let me see if I can manually integrate some of the suggestions I could't automatically add from the above.

marcoscaceres commented:
Ah, @TallTed  I see you made the same comment about GitHub confusion. 

Yeah, we can work out a better workflow with regards to recent changes... the problem is that I had to rebase after applying changes, which trashes/changes some of the history.  

Will try to avoid the rebase unless there would be a merge conflict.

mohamedamir commented on index.html:
```diff
@@ -539,7 +539,120 @@ <h2>
+    <aside class="note">
+      <p>
+        Although the coordinator handles input/output coordination, the
+        user-selected [=holder=] typically acts as the [=credential chooser=].
```
This isn't accurate for Android.
The platform presents the credential chooser not the wallet.

marcoscaceres replied:
Ok, I think it's the other way around on Apple's platforms - so we need to account for both. 

this is a good thing :) we've catered for the experience in different ways, but the result is identical.

marcoscaceres commented on index.html:
```diff
@@ -539,7 +539,120 @@ <h2>
+    <p>
+      To be written.
+    </p>
+    <h3>
+      Dismiss holder chooser
+    </h3>
+    <p>
+      To be written.
+    </p>
```
Ok, so, let's figure this out... what I'm trying to capture here is:

* what holds the abort controller and tells the UI to go away? 
* what/how do we tells the web application that the user has dismissed the UI? (then allowing new requests)

There needs to be something that coordinates the UI ceremony, as each step can result in an DOM exception, and settling the promise needs to happen at the appropriate time.

marcoscaceres commented on index.html:
```diff
@@ -539,7 +539,120 @@ <h2>
+    </ul>
+    <aside class="note">
+      <p>
+        Although the coordinator handles input/output coordination, the
+        user-selected [=holder=] typically acts as the [=credential chooser=].
```
```suggestion
        Although the coordinator handles input/output coordination,
        the user-selected [=holder=] acts as the [=credential chooser=]. Alternatively,
        the platform presents the [=credential chooser=] with credentials gathered from various [=holders=].
```

mohamedamir replied:
I would like to keep text at a more abstract level. User-selected holder hints that there is a holder selection screen is presented. I am proposing a more abstract text that leaves the details of coordination out of the spec.
What do you think?

```suggestion
Although the coordinator handles input/output coordination, it is the responsibility of the platform together with available [=holders=], to present the [=credential chooser=].
```

TallTed replied:
Tweaked to highlight @mohamedamir's changes. More tweaking will be required on the `together with available [=holders=]` phrase to make it proper English, if the meaning of this existing phrasing is accepted.
```suggestion
        Although the coordinator handles input/output coordination, it is the
        responsibility of the platform together with available [=holders=], to present the [=credential chooser=].
```
